app/assets/javascripts/visualizations/highvis/binaryTree.coffee

```diff
@@ -0,0 +1,393 @@
+
+      # Insert the BinaryTree object 'tree' at the location of the BinaryTree
+      # specified by index
+      ###
+      # WARNING:  MUTATES THE BINARY TREE 'THIS', DOES NOT MUTATE ARGUMENT TREE
+      ###
+      insertTree: (tree, index = 0) ->
+        replacementPoint = @index(index)
+        if index is null or replacementPoint is -1
+          console.log "Error inserting #{tree} at location specified.  Index does not exist in the tree."
+          null
+        else
+          start = replacementPoint.parent
+          if start isnt null
+            if start.left isnt null and window.BinaryTree.is_equal(replacementPoint, start.left)
+              start.left = window.BinaryTree.clone(tree)
+            else
+              if start.right isnt null and window.BinaryTree.is_equal(replacementPoint, start.right)
+                start.right = window.BinaryTree.clone(tree)
+            start.__updateParents()
+          else
+
+            @data = tree.data
+            @right = BinaryTree.clone(tree.right)
+            @left = BinaryTree.clone(tree.left)
+            @parent = null
+            @__updateParents()
+
+      # Updates binary tree element's parents to reflect the result of
+      # an insertTree merger to parent.right or parent.left
+      ###
+      # WARNING:  MUTATES THE BINARY TREE
+      ###
+      ###
+      # WARNING:  INTERNAL METHOD.  DO NOT CALL.
+      ###
+      __updateParents: ->
+        if @right isnt null
+          @right = window.BinaryTree.clone(@right, @)
+          @right.__updateParents()
+        if @left isnt null
+          @left = window.BinaryTree.clone(@left, @)
+          @left.__updateParents()
+
+      # Given two parent trees, create two new child trees by crossover.
+      # Both parent trees are given a randomly-selected crossover point.
+      # The first child is the part of the first parent before its crossover
+      # point, and the section of the second parent after its crossover point.
+      # The second child is the part of the first parent after its crossover
+      # point, and the part of the second parent before its crossover point.
+
+      @crossover: (tree1, tree2) ->
+        [tree1a, tree1b] = [window.BinaryTree.clone(tree1), window.BinaryTree.clone(tree1)]
+        [tree2a, tree2b] = [window.BinaryTree.clone(tree2), window.BinaryTree.clone(tree2)]
+        
+        [crossoverPointOne, crossoverPointTwo] = \
+        [Math.floor(Math.random() * tree1.treeSize()), Math.floor(Math.random() * tree2.treeSize())]
+        [childOne, childTwo] = [window.BinaryTree.clone(tree1a), window.BinaryTree.clone(tree2a)]
+        childOne.insertTree(childTwo.index(crossoverPointTwo), crossoverPointOne)
+        childTwo.insertTree(tree1b.index(crossoverPointOne), crossoverPointTwo)
+        [childOne, childTwo]
+
+      # Given a tree, construct a string representation of the mathematical
+      # function the tree describes
+      @stringify: (tree) ->
```

Comment:
Just a note here.

@stringify expands to this.stringify which is really window.stringify which is why this works.